### PR TITLE
Accommodate tunneling proxies that close the connection after an auth challenge.

### PR DESCRIPTION
**Pull request by dave-r12**
A fix for #2426. This was mainly a rejiggering of the existing connection code.

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/internal/io/RealConnection.java [hunk 1]
```diff
@@ -132,9 +133,52 @@ public void connect(int connectTimeout, int readTimeout, int writeTimeout,
...
+      }
+
+      connectSocket(connectTimeout, readTimeout, writeTimeout, connectionSpecSelector);
+      tunnelRequest = createTunnel(readTimeout, writeTimeout, tunnelRequest, url);
+
+      if (tunnelRequest == null) break; // Tunnel successfully created.
+
+      // The proxy decided to close the connection after an auth challenge. We need to create a new
+      // connection, but this time with the auth credentials.
+      closeQuietly(rawSocket);
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/io/RealConnection.java]: nit: also set rawSocket to null?

**Comment by swankjesse** [comment]
LGTM

This is brilliant. Very nice.

File: okhttp/src/main/java/okhttp3/internal/io/RealConnection.java (hunk 1 above)
Comment (dave-r12) [inline okhttp/src/main/java/okhttp3/internal/io/RealConnection.java]: Yes, done.

**Comment by axel3rd** [comment]
@dave-r12 / @swankjesse : Many thanks to you guys  :satisfied:

**Pull request merged**